pkg/scheduler/framework/v1alpha1/waiting_pods_map.go

```diff
@@ -69,26 +71,59 @@ func (m *waitingPodsMap) iterate(callback func(WaitingPod)) {
 
 // waitingPod represents a pod waiting in the permit phase.
 type waitingPod struct {
-	pod *v1.Pod
-	s   chan *Status
+	pod            *v1.Pod
+	pendingPlugins map[string]*time.Timer
+	s              chan *Status
 }
 
 // newWaitingPod returns a new waitingPod instance.
-func newWaitingPod(pod *v1.Pod) *waitingPod {
-	return &waitingPod{
+func newWaitingPod(pod *v1.Pod, pluginsMaxWaitTime map[string]time.Duration) *waitingPod {
+	wp := &waitingPod{
 		pod: pod,
 		s:   make(chan *Status),
 	}
+
+	pendingPlugins := make(map[string]*time.Timer)
+	for k, v := range pluginsMaxWaitTime {
+		plugin, waitTime := k, v
+		pendingPlugins[plugin] = time.AfterFunc(waitTime, func() {
+			msg := fmt.Sprintf("rejected due to timeout after waiting %v at plugin %v",
+				waitTime, plugin)
+			wp.Reject(msg)
+		})
+	}
```

Comment:
Have you tested that `plugin` and `waitTime` are being passed correctly? I think there may be some scope issues where all the `AfterFunc` called are using the last `waitTime` and `plugin` as they are being run in a loop and I think the fucntion is not being evaluated until the loop ends. You may need to use a wrapper function that receives as a param `plugin` and `waitTime` and returns a `func()` so that `plugin` and `waitTime` are evaluated during the loop.

Reply:
Line 88 solves the problem, they are local variables.

Reply:
Oh, didn't notice that trick. I usually go the wrapper function way.

Reply:
It is also a common solution :)